proposals/032-unique-generic-interface-parameter.md

```diff
@@ -0,0 +1,172 @@
+  * **Cons:** This approach significantly degrades ergonomics. A function signature that was once simple and clear:
+    ```swift
+    void processIntBuffer<B : IBuffer<int>>(B buffer);
+    ```
+    becomes much more verbose and less intuitive:
+    ```swift
+    void processIntBuffer<B : IBufferWithAssociatedType>(B buffer)
+        where B.Element == int;
+    ```
+    While functionally equivalent, this syntax is more burdensome for developers and harms readability. The `unique` proposal provides the same safety guarantees with a much better user experience.
+
+### Higher-Kinded Types
+
+The problem space has some overlap with higher-kinded types. A more advanced generic system could potentially offer alternative ways to abstract over "buffer-like" types. However, this is a much larger and more complex feature that would require significant design and implementation effort. Furthermore, it is not clear that it would cleanly solve this specific extension ambiguity problem without introducing its own complexities. The `unique` keyword provides a targeted, simple, and effective solution to the problem at hand.
+
+### Keyword Naming
+    The keyword `unique` is proposed as it clearly communicates the intent. However, the final name is open to discussion. An alternative like `associated` was briefly considered to hint at the link to associated types, but `unique` was deemed more direct.
+
+## Future Directions and Unresolved Questions
+
+### Deprecating Free-Form Generic `extension`
+
+The discussion that led to this proposal also highlighted a broader issue with the current `extension` mechanism. While this proposal makes `extension<T, U:IFoo<T>>` safer by introducing `unique` parameters, it is worth considering another powerful, and potentially problematic, form of extension: free-form generic `extension.
+
+This pattern takes the form:
+
+```slang
+extension<T : IFoo> T : IBar { /* ... */ }
+```
+
+This declaration states: "For any type `T` that conforms to `IFoo`, automatically make `T` also conform to `IBar`." While this can be a tool for reducing boilerplate, it introduces several significant problems that complicate the language and make code harder to reason about. This is particularly problematic with checking of `unique` keyword introduced in this proposal.
+
+For example, with this code:
+```slang
+interface IFoo<T> {}
+interface IBar {}
+struct Impl : IFoo<int>, IBar {}
+
+// this extension will make Impl conform to two different IFoo!
+extension<T:IBar> T: IFoo<float> {} 
+```
+It will be difficult for the compiler to find out who to blame for the duplication.
+
+The free form extension feature was originally implemented as a workaround to absence of default interface method implementations. Now that Slang is now supporting default interface methods with [SP #030](030-interface-method-default-impl.md), this main use of free-form generic extensions now has a better substitute and should be deprecated instead.
```

Comment:
I was initially hopeful that this proposal may allow my specific use case:
```slang
public extension<U : IHashable, T : IArray<U>> T: IHashable
{
    public uint64_t hash()
    {
        uint64_t h = 0;
        for (int i = 0; i < this.getCount(); ++i)
            h = hashCombine(h, this[i].hash());
        return h;
    }
}
```
But reading this, it seems that this kind of usage would instead be fully deprecated?